drivers/gpio/gpio_silabs_siwx91x.c

```diff
@@ -127,86 +106,147 @@ static int gpio_siwx91x_pin_configure(const struct device *dev, gpio_pin_t pin,
 		sl_si91x_gpio_select_ulp_pad_driver_disable_state(pin, disable_state);
 	} else {
-		sl_si91x_gpio_select_pad_driver_disable_state((cfg->port << 4) | pin,
+		sl_si91x_gpio_select_pad_driver_disable_state((port_cfg->port << 4) | pin,
 							      disable_state);
 	}
 
 	if (flags & GPIO_INPUT) {
-		if (cfg->ulp) {
+		if (port_cfg->ulp) {
 			sl_si91x_gpio_driver_enable_ulp_pad_receiver(pin);
 		} else {
-			sl_si91x_gpio_driver_enable_pad_receiver((cfg->port << 4) | pin);
+			sl_si91x_gpio_driver_enable_pad_receiver((port_cfg->port << 4) | pin);
 		}
 	} else {
-		if (cfg->ulp) {
+		if (port_cfg->ulp) {
 			sl_si91x_gpio_driver_disable_ulp_pad_receiver(pin);
 		} else {
-			sl_si91x_gpio_driver_disable_pad_receiver((cfg->port << 4) | pin);
+			sl_si91x_gpio_driver_disable_pad_receiver((port_cfg->port << 4) | pin);
 		}
 	}
 
-	pcfg->reg->PIN_CONFIG[(cfg->port << 4) + pin].GPIO_CONFIG_REG_b.MODE = 0;
+	cfg->reg->PIN_CONFIG[(port_cfg->port << 4) + pin].GPIO_CONFIG_REG_b.MODE = 0;
 
 	if (flags & GPIO_OUTPUT_INIT_HIGH) {
-		sl_gpio_set_pin_output(cfg->hal_port, pin);
+		sl_gpio_set_pin_output(port_cfg->hal_port, pin);
 	} else if (flags & GPIO_OUTPUT_INIT_LOW) {
-		sl_gpio_clear_pin_output(cfg->hal_port, pin);
+		sl_gpio_clear_pin_output(port_cfg->hal_port, pin);
+	}
+
+	sl_si91x_gpio_set_pin_direction(port_cfg->hal_port, pin, (flags & GPIO_OUTPUT) ? 0 : 1);
+
+	for (i = 0; i < port_data->next_pin; i++) {
+		if (port_data->pin_config_info[i].pin == pin) {
+			cur_cfg_pin = i;
+			break;
+		}
 	}
 
-	sl_si91x_gpio_set_pin_direction(cfg->hal_port, pin, (flags & GPIO_OUTPUT) ? 0 : 1);
+	if (i == port_data->next_pin) {
+		cur_cfg_pin = port_data->next_pin;
+		port_data->next_pin++;
+	}
 
+	if (cur_cfg_pin < __builtin_popcount(port_cfg->common.port_pin_mask)) {
+		port_data->pin_config_info[cur_cfg_pin].port_dev = dev;
+		port_data->pin_config_info[cur_cfg_pin].pin = pin;
+		port_data->pin_config_info[cur_cfg_pin].flags = flags;
+	} else {
+		return -EIO;
+	}
+
+	return 0;
+}
+
+static int gpio_siwx91x_pm_action(const struct device *dev, enum pm_device_action action)
+{
+	const struct gpio_siwx91x_common_config *cfg = dev->config;
+	struct gpio_siwx91x_common_data *data = dev->data;
+	const struct device **port_dev = data->ports;
+	const struct gpio_siwx91x_port_config *port_cfg = NULL;
+	struct gpio_siwx91x_port_data *port_data = NULL;
+	int pin_cnt;
+	int ret;
+
+	switch (action) {
+	case PM_DEVICE_ACTION_TURN_ON:
+		for (int i = 0; i < cfg->port_count; i++) {
+			port_cfg = port_dev[i]->config;
+			port_data = port_dev[i]->data;
+			pin_cnt = 0;
+			while (port_data->pin_config_info[pin_cnt].port_dev != NULL &&
+			       pin_cnt < __builtin_popcount(port_cfg->common.port_pin_mask)) {
+				ret = gpio_siwx91x_pin_configure(
+					port_data->pin_config_info[pin_cnt].port_dev,
+					port_data->pin_config_info[pin_cnt].pin,
+					port_data->pin_config_info[pin_cnt].flags);
+				if (ret) {
+					return ret;
+				}
+				pin_cnt++;
+			}
```

Comment:
can't you use the fact that you knows how much pin has been configured with your ```next_pin``` variable in port data ?

```suggestion
			for(int i = 0; i < port_data->next_pin; i++) {
				ret = gpio_siwx91x_pin_configure(
					port_data->pin_config_info[i].port_dev,
					port_data->pin_config_info[i].pin,
					port_data->pin_config_info[i].flags);
				if (ret) {
					return ret;
				}
			}
```

or something like this ? renaming to "pin_ctn" make more sense too here.